explorer/interpreter/type_checker.cpp

```diff
@@ -2360,6 +2360,21 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
           e->set_static_type(TupleValue::Empty());
           e->set_value_category(ValueCategory::Let);
           return Success();
+        case IntrinsicExpression::Intrinsic::Assert: {
+          if (args.size() != 2) {
+            return CompilationError(e->source_loc())
+                   << "__intrinsic_assert takes 2 arguments";
+          }
+          CARBON_RETURN_IF_ERROR(ExpectExactType(
+              e->source_loc(), "__intrinsic_assert argument 0",
+              arena_->New<BoolType>(), &args[0]->static_type(), impl_scope));
+          CARBON_RETURN_IF_ERROR(ExpectExactType(
```

Comment:
We shouldn't need to require an exact type match here, since the interpreter applies implicit conversion to this argument, so `ExpectType` should work.

It would be nice to allow implicit conversions on the first argument as well. That would mean switching it to use `ExpectType` here, and adding a suitable `Convert` call in the interpreter.

Reply:
Ok! Do you have an example which form of the call of __intrinsic_assert does not work now?

Reply:
I don't know of any types that are convertible to `bool` or `String` yet, so I can't easily concoct an example. I think we probably don't need to test this case yet, if that's what you're worried about; we just need to make the intent clear in the code.

Reply:
Done